### PR TITLE
Make MusicBrainzAlbumProvider thread safe and fix retry logic

### PR DESCRIPTION
**Pull request by Spacetech**
Related to #4242, and similar to #4223. 
Fixes issues when refreshing multiple items at the same time for musicbrainz. Unfortunately due to musicbrainz api limits, it seems like we need to keep this running sequentially.

Also fixes the retry logic - it's currently broken and throws this error whenever it tries the same request again:

```
System.InvalidOperationException: The request message was already sent. Cannot send the same request message multiple times.
   at System.Net.Http.HttpClient.CheckRequestMessage(HttpRequestMessage request)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request)
   at MediaBrowser.Providers.Music.MusicBrainzAlbumProvider.GetMusicBrainzResponse(String url, CancellationToken cancellationToken) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\MusicBrainz\MusicBrainzAlbumProvider.cs:line 774
   at MediaBrowser.Providers.Music.MusicBrainzAlbumProvider.GetReleaseResult(String albumName, String artistId, CancellationToken cancellationToken) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\MusicBrainz\MusicBrainzAlbumProvider.cs:line 285
   at MediaBrowser.Providers.Music.MusicBrainzAlbumProvider.GetMetadata(AlbumInfo id, CancellationToken cancellationToken) in C:\Programming\jellyfin\MediaBrowser.Providers\Plugins\MusicBrainz\MusicBrainzAlbumProvider.cs:line 213
   at MediaBrowser.Providers.Manager.MetadataService`2.ExecuteRemoteProviders(MetadataResult`1 temp, String logName, TIdType id, IEnumerable`1 providers, CancellationToken cancellationToken) in C:\Programming\jellyfin\MediaBrowser.Providers\Manager\MetadataService.cs:line 853
```

**Changes**
- Add a semaphore to only allow a single request at a time.
- Fix music brainz retry logic.

**Issues**
None

### CODE REVIEW

**Comment by dkanada** [comment]
Since you're having to make changes within the providers themselves, it sounds like even the non-core providers will need thread safety before we can parallelize the scanner.

**Comment by Spacetech** [comment]
> 
> 
> Since you're having to make changes within the providers themselves, it sounds like even the non-core providers will need thread safety before we can parallelize the scanner.

Yea I agree. I have been testing different types of libraries with different providers enabled in an attempt to find these issues. The TvDb provider and this one had obvious issues. I am somewhat worried there are issues in others too, but they might not be as visible (maybe they are rare race conditions).

Any suggestions for how to roll out the parallel scanner change? I'm not certain I will be able to find all these issues myself. If we are concerned about this, maybe we could merge the parallelized scanner change with the default concurrency set to 1, so the scanner would essentially remain sequential. Then users could change the config value to make it parallel and report back if they find issues?
I would certainly attempt to fix issues in any providers that are not working correctly with the parallelized scanner.

**Review by barronpm — APPROVED** [review]

File: MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs [hunk 1]
```diff
@@ -742,48 +743,58 @@ private string GetFirstReleaseGroupId(XmlReader reader)
...
-                }
+                    attempts++;
 
-                // Write time since last request to debug log as evidence we're meeting rate limit
-                // requirement, before resetting stopwatch back to zero.
-                _logger.LogDebug("GetMusicBrainzResponse: Time since previous request: {0} ms", _stopWatchMusicBrainz.ElapsedMilliseconds);
-                _stopWatchMusicBrainz.Restart();
+                    if (_stopWatchMusicBrainz.ElapsedMilliseconds < _musicBrainzQueryIntervalMs)
+                    {
+                        // MusicBrainz is extremely adamant about limiting to one request per second
```
Comment (BaronGreenback) [inline MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:760]:
```suggestion
                        // MusicBrainz is extremely adamant about limiting to one request per second.
```

File: MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs [hunk 2]
```diff
@@ -742,48 +743,58 @@ private string GetFirstReleaseGroupId(XmlReader reader)
...
+                        CultureInfo.InvariantCulture,
+                        "{0} ( {1} )",
+                        _appHost.ApplicationUserAgent,
+                        _appHost.ApplicationUserAgentAddress));
 
-            // Log error if unable to query MB database due to throttling
-            if (attempts == MusicBrainzQueryAttempts && response.StatusCode == HttpStatusCode.ServiceUnavailable)
+                    response = await _httpClientFactory.CreateClient(NamedClient.Default).SendAsync(request).ConfigureAwait(false);
+
+                    // We retry a finite number of times, and only whilst MB is indicating 503 (throttling)
```
Comment (BaronGreenback) [inline MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:782]:
```suggestion
                    // We retry a finite number of times, and only whilst MB is indicating 503 (throttling).
```

File: MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs [hunk 3]
```diff
@@ -742,48 +743,58 @@ private string GetFirstReleaseGroupId(XmlReader reader)
...
+                    _logger.LogDebug("GetMusicBrainzResponse: Time since previous request: {0} ms", _stopWatchMusicBrainz.ElapsedMilliseconds);
+                    _stopWatchMusicBrainz.Restart();
 
-                // We retry a finite number of times, and only whilst MB is indicating 503 (throttling)
-            }
-            while (attempts < MusicBrainzQueryAttempts && response.StatusCode == HttpStatusCode.ServiceUnavailable);
+                    using var request = new HttpRequestMessage(HttpMethod.Get, requestUrl);
+
+                    // MusicBrainz request a contact email address is supplied, as comment, in user agent field:
+                    // https://musicbrainz.org/doc/XML_Web_Service/Rate_Limiting#User-Agent
```
Comment (BaronGreenback) [inline MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:773]:
```suggestion
                    // https://musicbrainz.org/doc/XML_Web_Service/Rate_Limiting#User-Agent .
```

File: MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs [hunk 4]
```diff
@@ -742,48 +743,58 @@ private string GetFirstReleaseGroupId(XmlReader reader)
...
 
-            // Log error if unable to query MB database due to throttling
-            if (attempts == MusicBrainzQueryAttempts && response.StatusCode == HttpStatusCode.ServiceUnavailable)
+                    response = await _httpClientFactory.CreateClient(NamedClient.Default).SendAsync(request).ConfigureAwait(false);
+
+                    // We retry a finite number of times, and only whilst MB is indicating 503 (throttling)
+                }
+                while (attempts < MusicBrainzQueryAttempts && response.StatusCode == HttpStatusCode.ServiceUnavailable);
+
+                // Log error if unable to query MB database due to throttling
```
Comment (BaronGreenback) [inline MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:786]:
```suggestion
                // Log error if unable to query MB database due to throttling.
```

**Review by BaronGreenback — CHANGES_REQUESTED** [review]

**Review by BaronGreenback — APPROVED** [review]

**Pull request merged**